---
title: HW2 - Introduction to Biocluster and Linux
linkTitle: "HW2"
description: >
type: docs
weight: 302
---

## Topic: Linux Basics

1. Download code from this page
    ```sh
    wget https://cluster.hpcc.ucr.edu/~tgirke/Linux.sh --no-check-certificate 
    ```

2. Download Halobacterium proteome and inspect it
    ```sh
    wget https://ftp.ncbi.nlm.nih.gov/genomes/genbank/archaea/Halobacterium_salinarum/representative/GCA_004799605.1_ASM479960v1/GCA_004799605.1_ASM479960v1_protein.faa.gz
    gunzip GCA_004799605.1_ASM479960v1_protein.faa.gz
    mv GCA_004799605.1_ASM479960v1_protein.faa halobacterium.faa
    less halobacterium.faa # press q to quit
    ```

3. How many protein sequences are stored in the downloaded file?
    ```sh
    grep '>' halobacterium.faa | wc
    grep '^>' halobacterium.faa --count
    ```

4. How many proteins contain the pattern `WxHxxH` or `WxHxxHH`?
    ```sh
    egrep 'W.H..H{1,2}' halobacterium.faa --count
    ```

5. Use `less` to find IDs for pattern matches or use `awk`
    ```sh
    awk --posix -v RS='>' '/W.H..(H){1,2}/ { print ">" $0;}' halobacterium.faa | less
    awk --posix -v RS='>' '/W.H..(H){1,2}/ { print ">" $0;}' halobacterium.faa | grep '^>' | cut -c 2- | cut -f 1 -d\ > myIDs
    ```

6. Create a BLASTable database with `formatdb`
    ```sh
    module load ncbi-blast
    makeblastdb -in halobacterium.faa -out halobacterium.faa -dbtype prot -hash_index -parse_seqids
    ```

7. Query BLASTable database by IDs stored in a file (_e.g._ `myIDs`)
    ```sh
    blastdbcmd -db halobacterium.faa -dbtype prot -entry_batch myIDs -get_dups -out myseq.fasta
    ```

8. Run BLAST search for sequences stored in `myseq.fasta`
    ```sh
    blastp -query myseq.fasta -db halobacterium.faa -outfmt 0 -evalue 1e-6 -out blastp.out
    blastp -query myseq.fasta -db halobacterium.faa -outfmt 6 -evalue 1e-6 -out blastp.tab
    ```

Additional exercise material in [Linux Manual](https://hpcc.ucr.edu/manuals_linux-basics_shell.html)

## Homework assignment

Perform above analysis on the following _Escherichia coli_ strain: [Esch_coli_UMEA_3592-1_V1](ftp://ftp.ncbi.nih.gov/genomes/genbank/bacteria/Escherichia_coli/latest_assembly_versions/GCA_000461395.1_Esch_coli_UMEA_3592-1_V1/GCA_000461395.1_Esch_coli_UMEA_3592-1_V1_protein.faa.gz). 
Record result from final BLAST command (with `outfmt 6`) in text file.

## Homework submission

[GEN242-2021 HW2](https://classroom.github.com/a/HZMNsG_T) on GitHub Classroom.

In case you made some mistakes in the final file, to have the auto-grading to give you partial credits:

1. Upload your script and name it `hw2.sh`.
2. Upload the unzipped faa file from step 1, name it `ecoli.faa`.
3. Upload IDs from step 5 in a file named `myIDs`.
4. Upload the final file from step 8, name it `ecoli.txt`. 

If you are confident enough, only uploading the last file with the right answer will 
grant you full credits. 

## Due date

Most homeworks will be due one week after they are assigned. This one is due on Thu, April 15th at 6:00 PM.
