---
title: "Programming in R" 
author: "Author: Thomas Girke"
date: "Last update: 14 April, 2021" 
output:
  html_document:
    toc: true
    toc_float:
        collapsed: true
        smooth_scroll: true
    toc_depth: 3
    fig_caption: yes
    code_folding: show
    number_sections: true

fontsize: 14pt
bibliography: bibtex.bib
weight: 6
type: docs
---

<script src="/rmarkdown-libs/header-attrs/header-attrs.js"></script>


<!--
- Compile from command-line
Rscript -e "rmarkdown::render('Programming_in_R.Rmd', c('html_document'), clean=F); knitr::knit('Programming_in_R.Rmd', tangle=TRUE)"; Rscript ../md2jekyll.R Programming_in_R.knit.md 9; Rscript -e "rmarkdown::render('Programming_in_R.Rmd', c('pdf_document'))"
-->
<script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  document.querySelector("h1").className = "title";
});
</script>
<script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  var links = document.links;  
  for (var i = 0, linksLength = links.length; i < linksLength; i++)
    if (links[i].hostname != window.location.hostname)
      links[i].target = '_blank';
});
</script>
<div style="text-align: right">
<p>Source code downloads:    
[ <a href="https://raw.githubusercontent.com/tgirke/GEN242//main/content/en/tutorials/rprogramming/rprogramming.Rmd">.Rmd</a> ]    
[ <a href="https://raw.githubusercontent.com/tgirke/GEN242//main/content/en/tutorials/rprogramming/rprogramming.R">.R</a> ]</p>
</div>
<div id="overview" class="section level2">
<h2>Overview</h2>
<p>One of the main attractions of using the R
(<a href="http://cran.at.r-project.org">http://cran.at.r-project.org</a>) environment is
the ease with which users can write their own programs and custom functions.
The R programming syntax is extremely easy to learn, even for users with no
previous programming experience. Once the basic R programming control
structures are understood, users can use the R language as a powerful
environment to perform complex custom analyses of almost any type of data <span class="citation">(<a href="#ref-Gentleman2008-xo" role="doc-biblioref">Gentleman 2008</a>)</span>.</p>
</div>
<div id="why-programming-in-r" class="section level2">
<h2>Why Programming in R?</h2>
<ul>
<li>Powerful statistical environment and programming language</li>
<li>Facilitates reproducible research</li>
<li>Efficient data structures make programming very easy</li>
<li>Ease of implementing custom functions</li>
<li>Powerful graphics</li>
<li>Access to fast growing number of analysis packages</li>
<li>One of the most widely used languages in bioinformatics</li>
<li>Is standard for data mining and biostatistical analysis</li>
<li>Technical advantages: free, open-source, available for all OSs</li>
</ul>
</div>
<div id="r-basics" class="section level2">
<h2>R Basics</h2>
<p>The previous <a href="https://girke.bioinformatics.ucr.edu/GEN242/tutorials/rbasics/rbasics/">Rbasics</a> tutorial provides a general introduction to the usage of the R environment and its basic command syntax.
More details can be found in the R &amp; BioConductor manual <a href="http://manuals.bioinformatics.ucr.edu/home/R_BioCondManual">here</a>.</p>
</div>
<div id="code-editors-for-r" class="section level2">
<h2>Code Editors for R</h2>
<p>Several excellent code editors are available that provide functionalities like R syntax highlighting, auto code indenting and utilities to send code/functions to the R console.</p>
<ul>
<li><a href="https://www.rstudio.com/products/rstudio/features/">RStudio</a>: GUI-based IDE for R</li>
<li><a href="https://girke.bioinformatics.ucr.edu/GEN242/tutorials/linux/linux/#nvim-r-tmux-essentials">Vim-R-Tmux</a>: R working environment based on vim and tmux</li>
<li><a href="http://www.xemacs.org/Download/index.html">Emacs</a> (<a href="http://ess.r-project.org/">ESS add-on package</a>)</li>
<li><a href="https://wiki.gnome.org/Apps/Gedit">gedit</a> and <a href="https://wiki.gnome.org/Apps/Gedit">Rgedit</a></li>
<li><a href="https://rkward.kde.org/">RKWard</a></li>
<li><a href="http://www.walware.de/goto/statet">Eclipse</a></li>
<li><a href="https://sourceforge.net/projects/tinn-r/files/Tinn-R%20portable/">Tinn-R</a></li>
<li><a href="https://sourceforge.net/projects/npptor/">Notepad++ (NppToR)</a></li>
</ul>
<center>
Programming in R using RStudio
</center>
<center>
<img title="R_Interfaces" src="../images/rstudio.png"/>
</center>
<center>
Programming in R using Vim or Emacs
</center>
<center>
<img title="vim-r" src="../images/vimR.png"/>
</center>
</div>
<div id="finding-help" class="section level2">
<h2>Finding Help</h2>
<p>Reference list on R programming (selection)</p>
<ul>
<li><a href="http://adv-r.had.co.nz/">Advanced R</a>, by Hadley Wickham</li>
<li><a href="http://master.bioconductor.org/help/publications/books/r-programming-for-bioinformatics/">R Programming for Bioinformatics</a>, by Robert Gentleman</li>
<li><a href="http://www.stats.ox.ac.uk/pub/MASS3/Sprog/">S Programming</a>, by W. N. Venables and B. D. Ripley</li>
<li><a href="http://www.amazon.com/Programming-Data-Language-Lecture-Economics/dp/0387985034">Programming with Data</a>, by John M. Chambers</li>
<li><a href="http://www1.maths.lth.se/help/R/">R Help</a> &amp; <a href="http://www1.maths.lth.se/help/R/RCC/">R Coding Conventions</a>, Henrik Bengtsson, Lund University</li>
<li><a href="http://zoonek2.free.fr/UNIX/48_R/02.html">Programming in R</a> (Vincent Zoonekynd)</li>
<li><a href="http://www2.warwick.ac.uk/fac/sci/moac/people/students/peter_cock/r">Peter’s R Programming Pages</a>, University of Warwick</li>
<li><a href="http://pj.freefaculty.org/R/statsRus.html">Rtips</a>, Paul Johnsson, University of Kansas</li>
<li><a href="http://heather.cs.ucdavis.edu/~matloff/r.html">R for Programmers</a>, Norm Matloff, UC Davis</li>
<li><a href="http://www.statistik.uni-dortmund.de/useR-2008/tutorials/useR2008introhighperfR.pdf">High-Performance R</a>, Dirk Eddelbuettel tutorial presented at useR-2008</li>
<li><a href="http://www.stat.harvard.edu/ccr2005/index.html">C/C++ level programming for R</a>, Gopi Goswami</li>
</ul>
</div>
<div id="control-structures" class="section level2">
<h2>Control Structures</h2>
<div id="important-operators" class="section level3">
<h3>Important Operators</h3>
<div id="comparison-operators" class="section level4">
<h4>Comparison operators</h4>
<ul>
<li><code>==</code> (equal)</li>
<li><code>!=</code> (not equal)</li>
<li><code>&gt;</code> (greater than)</li>
<li><code>&gt;=</code> (greater than or equal)</li>
<li><code>&lt;</code> (less than)</li>
<li><code>&lt;=</code> (less than or equal)</li>
</ul>
</div>
<div id="logical-operators" class="section level4">
<h4>Logical operators</h4>
<ul>
<li><code>&amp;</code> (and)</li>
<li><code>|</code> (or)</li>
<li><code>!</code> (not)</li>
</ul>
</div>
</div>
<div id="conditional-executions-if-statements" class="section level3">
<h3>Conditional Executions: <code>if</code> Statements</h3>
<p>An <code>if</code> statement operates on length-one logical vectors.</p>
<p><strong>Syntax</strong></p>
<pre class="r"><code>if(TRUE) { 
    statements_1 
} else { 
    statements_2 
}</code></pre>
<p><strong>Example</strong></p>
<pre class="r"><code>if(1==0) { 
    print(1) 
} else { 
    print(2) 
}</code></pre>
<pre><code>## [1] 2</code></pre>
</div>
<div id="conditional-executions-ifelse-statements" class="section level3">
<h3>Conditional Executions: <code>ifelse</code> Statements</h3>
<p>The <code>ifelse</code> statement operates on vectors.</p>
<p><strong>Syntax</strong></p>
<pre class="r"><code>ifelse(test, true_value, false_value)</code></pre>
<p><strong>Example</strong></p>
<pre class="r"><code>x &lt;- 1:10 
ifelse(x&lt;5, sqrt(x), 0)</code></pre>
<pre><code>##  [1] 1.000000 1.414214 1.732051 2.000000 0.000000 0.000000 0.000000 0.000000 0.000000 0.000000</code></pre>
</div>
</div>
<div id="loops" class="section level2">
<h2>Loops</h2>
<div id="for-loop" class="section level3">
<h3><code>for</code> loop</h3>
<p><code>for</code> loops iterate over elements of a looping vector.</p>
<p><strong>Syntax</strong></p>
<pre class="r"><code>for(variable in sequence) { 
    statements 
}</code></pre>
<p><strong>Example</strong></p>
<pre class="r"><code>mydf &lt;- iris
myve &lt;- NULL
for(i in seq(along=mydf[,1])) {
    myve &lt;- c(myve, mean(as.numeric(mydf[i,1:3])))
}
myve[1:8]</code></pre>
<pre><code>## [1] 3.333333 3.100000 3.066667 3.066667 3.333333 3.666667 3.133333 3.300000</code></pre>
<p><strong>Note:</strong> Inject into objecs is much faster than append approach with <code>c</code>, <code>cbind</code>, etc.</p>
<p><strong>Example</strong></p>
<pre class="r"><code>myve &lt;- numeric(length(mydf[,1]))
for(i in seq(along=myve)) {
    myve[i] &lt;- mean(as.numeric(mydf[i,1:3]))
}
myve[1:8]</code></pre>
<pre><code>## [1] 3.333333 3.100000 3.066667 3.066667 3.333333 3.666667 3.133333 3.300000</code></pre>
<div id="conditional-stop-of-loops" class="section level4">
<h4>Conditional Stop of Loops</h4>
<p>The <code>stop</code> function can be used to break out of a loop (or a function) when a condition becomes <code>TRUE</code>. In addition, an error message will be printed.</p>
<p><strong>Example</strong></p>
<pre class="r"><code>x &lt;- 1:10
z &lt;- NULL
for(i in seq(along=x)) { 
    if(x[i] &lt; 5) { 
        z &lt;- c(z, x[i]-1)  
    } else { 
        stop(&quot;values need to be &lt; 5&quot;) 
    }
}</code></pre>
</div>
</div>
<div id="while-loop" class="section level3">
<h3><code>while</code> loop</h3>
<p>Iterates as long as a condition is true.</p>
<p><strong>Syntax</strong></p>
<pre class="r"><code>while(condition) {
    statements
}</code></pre>
<p><strong>Example</strong></p>
<pre class="r"><code>z &lt;- 0
while(z&lt;5) { 
    z &lt;- z + 2
    print(z)  
}</code></pre>
<pre><code>## [1] 2
## [1] 4
## [1] 6</code></pre>
</div>
<div id="the-apply-function-family" class="section level3">
<h3>The <code>apply</code> Function Family</h3>
<div id="apply" class="section level4">
<h4><code>apply</code></h4>
<p><strong>Syntax</strong></p>
<pre class="r"><code>apply(X, MARGIN, FUN, ARGs)</code></pre>
<p><strong>Arguments</strong></p>
<ul>
<li><code>X</code>: <code>array</code>, <code>matrix</code> or <code>data.frame</code></li>
<li><code>MARGIN</code>: <code>1</code> for rows, <code>2</code> for columns</li>
<li><code>FUN</code>: one or more functions</li>
<li><code>ARGs</code>: possible arguments for functions</li>
</ul>
<p><strong>Example</strong></p>
<pre class="r"><code>apply(iris[1:8,1:3], 1, mean)</code></pre>
<pre><code>##        1        2        3        4        5        6        7        8 
## 3.333333 3.100000 3.066667 3.066667 3.333333 3.666667 3.133333 3.300000</code></pre>
</div>
<div id="tapply" class="section level4">
<h4><code>tapply</code></h4>
<p>Applies a function to vector components that are defined by a factor.</p>
<p><strong>Syntax</strong></p>
<pre class="r"><code>tapply(vector, factor, FUN)</code></pre>
<p><strong>Example</strong></p>
<pre class="r"><code>iris[1:2,]</code></pre>
<pre><code>##   Sepal.Length Sepal.Width Petal.Length Petal.Width Species
## 1          5.1         3.5          1.4         0.2  setosa
## 2          4.9         3.0          1.4         0.2  setosa</code></pre>
<pre class="r"><code>tapply(iris$Sepal.Length, iris$Species, mean)</code></pre>
<pre><code>##     setosa versicolor  virginica 
##      5.006      5.936      6.588</code></pre>
</div>
<div id="sapply-lapply-and-vapply" class="section level4">
<h4><code>sapply</code>, <code>lapply</code> and <code>vapply</code></h4>
<p>The iterator functions <code>sapply</code>, <code>lapply</code> and <code>vapply</code> apply a function to
vectors or lists. The <code>lapply</code> function always returns a list, while <code>sapply</code>
returns <code>vector</code> or <code>matrix</code> objects when possible. If not then a list is
returned. The <code>vapply</code> function returns a vector or array of type matching the
<code>FUN.VALUE</code>. Compared to <code>sapply</code>, <code>vapply</code> is a safer choice with respect to
controlling specific output types to avoid exception handling problems.</p>
<p><strong>Examples</strong></p>
<pre class="r"><code>l &lt;- list(a = 1:10, beta = exp(-3:3), logic = c(TRUE,FALSE,FALSE,TRUE))
lapply(l, mean)</code></pre>
<pre><code>## $a
## [1] 5.5
## 
## $beta
## [1] 4.535125
## 
## $logic
## [1] 0.5</code></pre>
<pre class="r"><code>sapply(l, mean)</code></pre>
<pre><code>##        a     beta    logic 
## 5.500000 4.535125 0.500000</code></pre>
<pre class="r"><code>vapply(l, mean, FUN.VALUE=numeric(1))</code></pre>
<pre><code>##        a     beta    logic 
## 5.500000 4.535125 0.500000</code></pre>
<p>Often used in combination with a function definition:</p>
<pre class="r"><code>lapply(names(l), function(x) mean(l[[x]]))
sapply(names(l), function(x) mean(l[[x]]))
vapply(names(l), function(x) mean(l[[x]]), FUN.VALUE=numeric(1))</code></pre>
</div>
</div>
</div>
<div id="functions" class="section level2">
<h2>Functions</h2>
<div id="function-overview" class="section level3">
<h3>Function Overview</h3>
<p>A very useful feature of the R environment is the possibility to expand existing functions and to easily write custom functions. In fact, most of the R software can be viewed as a series of R functions.</p>
<p><strong>Syntax</strong> to define function</p>
<pre class="r"><code>myfct &lt;- function(arg1, arg2, ...) { 
    function_body 
}</code></pre>
<p><strong>Syntax</strong> to call functions</p>
<pre class="r"><code>myfct(arg1=..., arg2=...)</code></pre>
<p>The value returned by a function is the value of the function body, which is usually an unassigned final expression, <em>e.g.</em>: <code>return()</code></p>
</div>
<div id="function-syntax-rules" class="section level3">
<h3>Function Syntax Rules</h3>
<p><strong>General</strong></p>
<ul>
<li>Functions are defined by
<ol style="list-style-type: decimal">
<li>The assignment with the keyword <code>function</code></li>
<li>The declaration of arguments/variables (<code>arg1, arg2, ...</code>)</li>
<li>The definition of operations (<code>function_body</code>) that perform computations on the provided arguments. A function name needs to be assigned to call the function.</li>
</ol></li>
</ul>
<p><strong>Naming</strong></p>
<ul>
<li>Function names can be almost anything. However, the usage of names of existing functions should be avoided.</li>
</ul>
<p><strong>Arguments</strong></p>
<ul>
<li>It is often useful to provide default values for arguments (<em>e.g.</em>: <code>arg1=1:10</code>). This way they don’t need to be provided in a function call. The argument list can also be left empty (<code>myfct &lt;- function() { fct_body }</code>) if a function is expected to return always the same value(s). The argument <code>...</code> can be used to allow one function to pass on argument settings to another.</li>
</ul>
<p><strong>Body</strong></p>
<ul>
<li>The actual expressions (commands/operations) are defined in the function body which should be enclosed by braces. The individual commands are separated by semicolons or new lines (preferred).</li>
</ul>
<p><strong>Usage</strong></p>
<ul>
<li>Functions are called by their name followed by parentheses containing possible argument names. Empty parenthesis after the function name will result in an error message when a function requires certain arguments to be provided by the user. The function name alone will print the definition of a function.</li>
</ul>
<p><strong>Scope</strong></p>
<ul>
<li>Variables created inside a function exist only for the life time of a function. Thus, they are not accessible outside of the function. To force variables in functions to exist globally, one can use the double assignment operator: <code>&lt;&lt;-</code></li>
</ul>
</div>
<div id="examples" class="section level3">
<h3>Examples</h3>
<p><strong>Define sample function</strong></p>
<pre class="r"><code>myfct &lt;- function(x1, x2=5) { 
    z1 &lt;- x1 / x1
    z2 &lt;- x2 * x2
        myvec &lt;- c(z1, z2) 
        return(myvec)
} </code></pre>
<p><strong>Function usage</strong></p>
<p>Apply function to values <code>2</code> and <code>5</code></p>
<pre class="r"><code>myfct(x1=2, x2=5) </code></pre>
<pre><code>## [1]  1 25</code></pre>
<p>Run without argument names</p>
<pre class="r"><code>myfct(2, 5) </code></pre>
<pre><code>## [1]  1 25</code></pre>
<p>Makes use of default value <code>5</code></p>
<pre class="r"><code>myfct(x1=2) </code></pre>
<pre><code>## [1]  1 25</code></pre>
<p>Print function definition (often unintended)</p>
<pre class="r"><code>myfct </code></pre>
<pre><code>## function(x1, x2=5) { 
##  z1 &lt;- x1 / x1
##  z2 &lt;- x2 * x2
##         myvec &lt;- c(z1, z2) 
##         return(myvec)
## }
## &lt;bytecode: 0x5865ccde4df0&gt;</code></pre>
</div>
</div>
<div id="useful-utilities" class="section level2">
<h2>Useful Utilities</h2>
<div id="debugging-utilities" class="section level3">
<h3>Debugging Utilities</h3>
<p>Several debugging utilities are available for R. They include:</p>
<ul>
<li><code>traceback</code></li>
<li><code>browser</code></li>
<li><code>options(error=recover)</code></li>
<li><code>options(error=NULL)</code></li>
<li><code>debug</code></li>
</ul>
<p>The <a href="https://adv-r.hadley.nz/debugging.html">Debugging in R page</a> provides an overview of the available resources.</p>
</div>
<div id="regular-expressions" class="section level3">
<h3>Regular Expressions</h3>
<p>R’s regular expression utilities work similar as in other languages. To learn how to use them in R, one can consult the main help page on this topic with <code>?regexp</code>.</p>
<div id="string-matching-with-grep" class="section level4">
<h4>String matching with <code>grep</code></h4>
<p>The grep function can be used for finding patterns in strings, here letter <code>A</code> in vector <code>month.name</code>.</p>
<pre class="r"><code>month.name[grep(&quot;A&quot;, month.name)] </code></pre>
<pre><code>## [1] &quot;April&quot;  &quot;August&quot;</code></pre>
</div>
<div id="string-substitution-with-gsub" class="section level4">
<h4>String substitution with <code>gsub</code></h4>
<p>Example for using regular expressions to substitute a pattern by another one using a back reference. Remember: single escapes <code>\</code> need to be double escaped <code>\\</code> in R.</p>
<pre class="r"><code>gsub(&#39;(i.*a)&#39;, &#39;xxx_\\1&#39;, &quot;virginica&quot;, perl = TRUE) </code></pre>
<pre><code>## [1] &quot;vxxx_irginica&quot;</code></pre>
</div>
</div>
<div id="interpreting-a-character-string-as-expression" class="section level3">
<h3>Interpreting a Character String as Expression</h3>
<p>Some useful examples</p>
<p>Generates vector of object names in session</p>
<pre class="r"><code>myfct &lt;- function(x) x^2
mylist &lt;- ls()
n &lt;- which(mylist %in% &quot;myfct&quot;)
mylist[n] </code></pre>
<pre><code>## [1] &quot;myfct&quot;</code></pre>
<p>Executes entry in position <code>n</code> as expression</p>
<pre class="r"><code>get(mylist[n])</code></pre>
<pre><code>## function(x) x^2</code></pre>
<pre class="r"><code>get(mylist[n])(2)</code></pre>
<pre><code>## [1] 4</code></pre>
<p>Alternative approach</p>
<pre class="r"><code>eval(parse(text=mylist[n])) </code></pre>
<pre><code>## function(x) x^2</code></pre>
</div>
<div id="replacement-split-and-paste-functions-for-strings" class="section level3">
<h3>Replacement, Split and Paste Functions for Strings</h3>
<p><strong>Selected examples</strong></p>
<p>Substitution with back reference which inserts in this example <code>_</code> character</p>
<pre class="r"><code>x &lt;- gsub(&quot;(a)&quot;,&quot;\\1_&quot;, month.name[1], perl=T) 
x</code></pre>
<pre><code>## [1] &quot;Ja_nua_ry&quot;</code></pre>
<p>Split string on inserted character from above</p>
<pre class="r"><code>strsplit(x,&quot;_&quot;)</code></pre>
<pre><code>## [[1]]
## [1] &quot;Ja&quot;  &quot;nua&quot; &quot;ry&quot;</code></pre>
<p>Reverse a character string by splitting first all characters into vector fields</p>
<pre class="r"><code>paste(rev(unlist(strsplit(x, NULL))), collapse=&quot;&quot;) </code></pre>
<pre><code>## [1] &quot;yr_aun_aJ&quot;</code></pre>
</div>
<div id="time-date-and-sleep" class="section level3">
<h3>Time, Date and Sleep</h3>
<p><strong>Selected examples</strong></p>
<p>Return CPU (and other) times that an expression used (here ls)</p>
<pre class="r"><code>system.time(ls()) </code></pre>
<pre><code>##    user  system elapsed 
##       0       0       0</code></pre>
<p>Return the current system date and time</p>
<pre class="r"><code>date() </code></pre>
<pre><code>## [1] &quot;Sat Feb 13 18:30:39 2021&quot;</code></pre>
<p>Pause execution of R expressions for a given number of seconds (e.g. in loop)</p>
<pre class="r"><code>Sys.sleep(1) </code></pre>
<div id="example" class="section level4">
<h4>Example</h4>
<div id="import-of-specific-file-lines-with-regular-expression" class="section level5">
<h5>Import of Specific File Lines with Regular Expression</h5>
<p>The following example demonstrates the retrieval of specific lines from an external file with a regular expression. First, an external file is created with the <code>cat</code> function, all lines of this file are imported into a vector with <code>readLines</code>, the specific elements (lines) are then retieved with the <code>grep</code> function, and the resulting lines are split into vector fields with <code>strsplit</code>.</p>
<pre class="r"><code>cat(month.name, file=&quot;zzz.txt&quot;, sep=&quot;\n&quot;)
x &lt;- readLines(&quot;zzz.txt&quot;)
x[1:6] </code></pre>
<pre><code>## [1] &quot;January&quot;  &quot;February&quot; &quot;March&quot;    &quot;April&quot;    &quot;May&quot;      &quot;June&quot;</code></pre>
<pre class="r"><code>x &lt;- x[c(grep(&quot;^J&quot;, as.character(x), perl = TRUE))]
t(as.data.frame(strsplit(x, &quot;u&quot;)))</code></pre>
<pre><code>##                 [,1]  [,2] 
## c..Jan....ary.. &quot;Jan&quot; &quot;ary&quot;
## c..J....ne..    &quot;J&quot;   &quot;ne&quot; 
## c..J....ly..    &quot;J&quot;   &quot;ly&quot;</code></pre>
</div>
</div>
</div>
</div>
<div id="calling-external-software" class="section level2">
<h2>Calling External Software</h2>
<p>External command-line software can be called with <code>system</code>. The following example calls <code>blastall</code> from R</p>
<pre class="r"><code>system(&quot;blastall -p blastp -i seq.fasta -d uniprot -o seq.blastp&quot;)</code></pre>
</div>
<div id="running-r-scripts" class="section level2">
<h2>Running R Scripts</h2>
<div id="possibilities-for-executing-r-scripts" class="section level3">
<h3>Possibilities for Executing R Scripts</h3>
<div id="r-console" class="section level4">
<h4>R console</h4>
<pre class="r"><code>source(&quot;my_script.R&quot;)</code></pre>
</div>
<div id="command-line" class="section level4">
<h4>Command-line</h4>
<pre class="sh"><code>Rscript my_script.R # or just ./myscript.R after making it executable
R CMD BATCH my_script.R # Alternative way 1 
R --slave &lt; my_script.R # Alternative way 2</code></pre>
</div>
<div id="passing-arguments-from-command-line-to-r" class="section level4">
<h4>Passing arguments from command-line to R</h4>
<p>Create an R script named <code>test.R</code> with the following content:</p>
<pre class="sh"><code>myarg &lt;- commandArgs()
print(iris[1:myarg[6], ])</code></pre>
<p>Then run it from the command-line like this:</p>
<pre class="sh"><code>Rscript test.R 10</code></pre>
<p>In the given example the number <code>10</code> is passed on from the command-line as an argument to the R script which is used to return to <code>STDOUT</code> the first 10 rows of the <code>iris</code> sample data. If several arguments are provided, they will be interpreted as one string and need to be split in R with the strsplit function. A more detailed example can be found <a href="http://manuals.bioinformatics.ucr.edu/home/ht-seq#TOC-Quality-Reports-of-FASTQ-Files-">here</a>.</p>
</div>
</div>
</div>
<div id="building-r-packages" class="section level2">
<h2>Building R Packages</h2>
<div id="short-overview-of-package-building-process" class="section level3">
<h3>Short Overview of Package Building Process</h3>
<p>R packages can be built with the <code>package.skeleton</code> function. The given example will create a directory named <code>mypackage</code> containing the skeleton of the package for all functions, methods and classes defined in the R script(s) passed on to the <code>code_files</code> argument. The basic structure of the package directory is described <a href="http://manuals.bioinformatics.ucr.edu/home/programming-in-r#Progr_pack">here</a>. The package directory will also contain a file named <code>Read-and-delete-me</code> with instructions for completing the package:</p>
<pre class="r"><code>xxxx  package.skeleton(name=&quot;mypackage&quot;, code_files=c(&quot;script1.R&quot;, &quot;script2.R&quot;))</code></pre>
<p>Once a package skeleton is available one can build the package from the command-line (Linux/OS X). This will create a tarball of the package with its version number encoded in the file name. Subequently, the package tarball needs to be checked for errors with:</p>
<pre class="r"><code>R CMD build mypackage
R CMD check mypackage_1.0.tar.gz</code></pre>
<p>Install package from source</p>
<pre class="r"><code>install.packages(&quot;mypackage_1.0.tar.gz&quot;, repos=NULL) </code></pre>
<p>For more details see <a href="http://manuals.bioinformatics.ucr.edu/home/programming-in-r#TOC-Building-R-Packages">here</a></p>
</div>
</div>
<div id="programming-exercises" class="section level2">
<h2>Programming Exercises</h2>
<div id="exercise-1" class="section level3">
<h3>Exercise 1</h3>
<div id="for-loop-1" class="section level4">
<h4><code>for</code> loop</h4>
<p><strong>Task 1.1</strong>: Compute the mean of each row in <code>myMA</code> by applying the mean function in a <code>for</code> loop.</p>
<pre class="r"><code>myMA &lt;- matrix(rnorm(500), 100, 5, dimnames=list(1:100, paste(&quot;C&quot;, 1:5, sep=&quot;&quot;)))
myve_for &lt;- NULL
for(i in seq(along=myMA[,1])) {
    myve_for &lt;- c(myve_for, mean(as.numeric(myMA[i, ])))
}
myResult &lt;- cbind(myMA, mean_for=myve_for)
myResult[1:4, ]</code></pre>
<pre><code>##           C1           C2         C3         C4         C5    mean_for
## 1  1.7988703  0.785864023  1.2763288  0.2950553  0.9858471  1.02839310
## 2 -0.9683694  2.724133446 -0.8274809 -0.3208423 -0.4600894  0.02947028
## 3 -1.1476365 -0.002864923  1.2573494  0.9574395  1.0390970  0.42067691
## 4 -0.3476123  0.639676778 -1.4706145 -1.3904486 -0.1612300 -0.54604574</code></pre>
</div>
<div id="while-loop-1" class="section level4">
<h4><code>while</code> loop</h4>
<p><strong>Task 1.2</strong>: Compute the mean of each row in <code>myMA</code> by applying the mean function in a <code>while</code> loop.</p>
<pre class="r"><code>z &lt;- 1
myve_while &lt;- NULL
while(z &lt;= length(myMA[,1])) {
    myve_while &lt;- c(myve_while, mean(as.numeric(myMA[z, ])))
    z &lt;- z + 1
}
myResult &lt;- cbind(myMA, mean_for=myve_for, mean_while=myve_while)
myResult[1:4, -c(1,2)]</code></pre>
<pre><code>##           C3         C4         C5    mean_for  mean_while
## 1  1.2763288  0.2950553  0.9858471  1.02839310  1.02839310
## 2 -0.8274809 -0.3208423 -0.4600894  0.02947028  0.02947028
## 3  1.2573494  0.9574395  1.0390970  0.42067691  0.42067691
## 4 -1.4706145 -1.3904486 -0.1612300 -0.54604574 -0.54604574</code></pre>
<p><strong>Task 1.3</strong>: Confirm that the results from both mean calculations are identical</p>
<pre class="r"><code>all(myResult[,6] == myResult[,7])</code></pre>
<pre><code>## [1] TRUE</code></pre>
</div>
<div id="apply-loop" class="section level4">
<h4><code>apply</code> loop</h4>
<p><strong>Task 1.4</strong>: Compute the mean of each row in myMA by applying the mean function in an <code>apply</code> loop</p>
<pre class="r"><code>myve_apply &lt;- apply(myMA, 1, mean)
myResult &lt;- cbind(myMA, mean_for=myve_for, mean_while=myve_while, mean_apply=myve_apply)
myResult[1:4, -c(1,2)]</code></pre>
<pre><code>##           C3         C4         C5    mean_for  mean_while  mean_apply
## 1  1.2763288  0.2950553  0.9858471  1.02839310  1.02839310  1.02839310
## 2 -0.8274809 -0.3208423 -0.4600894  0.02947028  0.02947028  0.02947028
## 3  1.2573494  0.9574395  1.0390970  0.42067691  0.42067691  0.42067691
## 4 -1.4706145 -1.3904486 -0.1612300 -0.54604574 -0.54604574 -0.54604574</code></pre>
</div>
<div id="avoiding-loops" class="section level4">
<h4>Avoiding loops</h4>
<p><strong>Task 1.5</strong>: When operating on large data sets it is much faster to use the rowMeans function</p>
<pre class="r"><code>mymean &lt;- rowMeans(myMA)
myResult &lt;- cbind(myMA, mean_for=myve_for, mean_while=myve_while, mean_apply=myve_apply, mean_int=mymean)
myResult[1:4, -c(1,2,3)]</code></pre>
<pre><code>##           C4         C5    mean_for  mean_while  mean_apply    mean_int
## 1  0.2950553  0.9858471  1.02839310  1.02839310  1.02839310  1.02839310
## 2 -0.3208423 -0.4600894  0.02947028  0.02947028  0.02947028  0.02947028
## 3  0.9574395  1.0390970  0.42067691  0.42067691  0.42067691  0.42067691
## 4 -1.3904486 -0.1612300 -0.54604574 -0.54604574 -0.54604574 -0.54604574</code></pre>
</div>
</div>
<div id="exercise-2" class="section level3">
<h3>Exercise 2</h3>
<div id="custom-functions" class="section level4">
<h4>Custom functions</h4>
<p><strong>Task 2.1</strong>: Use the following code as basis to implement a function that allows the user to compute the mean for any combination of columns in a matrix or data frame. The first argument of this function should specify the input data set, the second the mathematical function to be passed on (<em>e.g.</em> <code>mean</code>, <code>sd</code>, <code>max</code>) and the third one should allow the selection of the columns by providing a grouping vector.</p>
<pre class="r"><code>myMA &lt;- matrix(rnorm(100000), 10000, 10, dimnames=list(1:10000, paste(&quot;C&quot;, 1:10, sep=&quot;&quot;)))
myMA[1:2,]</code></pre>
<pre><code>##           C1         C2         C3         C4        C5         C6        C7          C8        C9
## 1  0.2457294  1.0377898 -0.1467638 -0.3248598 0.6157826 -1.4879414 2.0392123 -0.08362708 0.5044756
## 2 -0.8500019 -0.7396865 -0.2914722  1.0449035 0.4690041 -0.2100562 0.7917301  0.17614233 0.3136505
##          C10
## 1 -0.6699119
## 2 -0.6938574</code></pre>
<pre class="r"><code>myList &lt;- tapply(colnames(myMA), c(1,1,1,2,2,2,3,3,4,4), list) 
names(myList) &lt;- sapply(myList, paste, collapse=&quot;_&quot;)
myMAmean &lt;- sapply(myList, function(x) apply(myMA[,x], 1, mean))
myMAmean[1:4,] </code></pre>
<pre><code>##      C1_C2_C3    C4_C5_C6      C7_C8      C9_C10
## 1  0.37891844 -0.39900621  0.9777926 -0.08271812
## 2 -0.62705354  0.43461709  0.4839362 -0.19010345
## 3  0.44567197 -0.03322511 -0.1556891 -0.56857498
## 4 -0.07690464  0.66825754 -0.3142668 -0.43291652</code></pre>
<!---
Solution

-->
</div>
</div>
<div id="exercise-3" class="section level3">
<h3>Exercise 3</h3>
<div id="nested-loops-to-generate-similarity-matrices" class="section level4">
<h4>Nested loops to generate similarity matrices</h4>
<p><strong>Task 3.1</strong>: Create a sample list populated with character vectors of different lengths</p>
<pre class="r"><code>setlist &lt;- lapply(11:30, function(x) sample(letters, x, replace=TRUE))
names(setlist) &lt;- paste(&quot;S&quot;, seq(along=setlist), sep=&quot;&quot;) 
setlist[1:6]</code></pre>
<pre><code>## $S1
##  [1] &quot;t&quot; &quot;y&quot; &quot;l&quot; &quot;w&quot; &quot;x&quot; &quot;c&quot; &quot;w&quot; &quot;i&quot; &quot;n&quot; &quot;v&quot; &quot;v&quot;
## 
## $S2
##  [1] &quot;t&quot; &quot;f&quot; &quot;q&quot; &quot;z&quot; &quot;e&quot; &quot;g&quot; &quot;u&quot; &quot;m&quot; &quot;q&quot; &quot;t&quot; &quot;l&quot; &quot;f&quot;
## 
## $S3
##  [1] &quot;j&quot; &quot;g&quot; &quot;t&quot; &quot;z&quot; &quot;l&quot; &quot;q&quot; &quot;p&quot; &quot;k&quot; &quot;a&quot; &quot;k&quot; &quot;s&quot; &quot;q&quot; &quot;y&quot;
## 
## $S4
##  [1] &quot;y&quot; &quot;g&quot; &quot;j&quot; &quot;n&quot; &quot;a&quot; &quot;x&quot; &quot;h&quot; &quot;a&quot; &quot;t&quot; &quot;q&quot; &quot;f&quot; &quot;u&quot; &quot;p&quot; &quot;n&quot;
## 
## $S5
##  [1] &quot;i&quot; &quot;c&quot; &quot;y&quot; &quot;e&quot; &quot;k&quot; &quot;u&quot; &quot;j&quot; &quot;v&quot; &quot;t&quot; &quot;u&quot; &quot;q&quot; &quot;w&quot; &quot;b&quot; &quot;q&quot; &quot;s&quot;
## 
## $S6
##  [1] &quot;w&quot; &quot;i&quot; &quot;d&quot; &quot;u&quot; &quot;w&quot; &quot;m&quot; &quot;o&quot; &quot;u&quot; &quot;v&quot; &quot;k&quot; &quot;m&quot; &quot;t&quot; &quot;b&quot; &quot;h&quot; &quot;h&quot; &quot;v&quot;</code></pre>
<p><strong>Task 3.2</strong>: Compute the length for all pairwise intersects of the vectors stored in <code>setlist</code>. The intersects can be determined with the <code>%in%</code> function like this: <code>sum(setlist[[1]] %in% setlist[[2]])</code></p>
<pre class="r"><code>setlist &lt;- sapply(setlist, unique)
olMA &lt;- sapply(names(setlist), function(x) sapply(names(setlist), 
               function(y) sum(setlist[[x]] %in% setlist[[y]])))
olMA[1:12,] </code></pre>
<pre><code>##     S1 S2 S3 S4 S5 S6 S7 S8 S9 S10 S11 S12 S13 S14 S15 S16 S17 S18 S19 S20
## S1   9  2  3  4  6  4  4  4  5   5   6   5   5   3   5   5   4   6   7   7
## S2   2  9  5  5  4  3  4  5  4   6   4   6   5   7   5   4   5   7   5   7
## S3   3  5 11  7  6  2  4  6  5   3   7   7   5   8   6   5   7   5   6  10
## S4   4  5  7 12  5  3  6  6  3   6   5   6   6   8   7   6   8   6   7   9
## S5   6  4  6  5 13  7  4  3  5   6   9   7   7   7   9   4   8   8   8  11
## S6   4  3  2  3  7 11  3  3  4   7   6   8   4   6   6   6   8   9   7   8
## S7   4  4  4  6  4  3 12  6  5   7   6   6   8   7   7   7   9   8   9   9
## S8   4  5  6  6  3  3  6 10  7   5   5   5   6   7   4   7   6   5   7   9
## S9   5  4  5  3  5  4  5  7 11   5   6   7   6   6   7   7   5   6   7   9
## S10  5  6  3  6  6  7  7  5  5  14   6   8   6   8   7   9   8   8  11   9
## S11  6  4  7  5  9  6  6  5  6   6  13   7   4   8  10   6  10   9  10  10
## S12  5  6  7  6  7  8  6  5  7   8   7  15   6  10   8   9  10  10   7  11</code></pre>
<p><strong>Task 3.3</strong> Plot the resulting intersect matrix as heat map. The <code>image</code> or the <code>heatmap.2</code> function from the <code>gplots</code> library can be used for this.</p>
<pre class="r"><code>image(olMA)</code></pre>
<p><img src="/en/tutorials/rprogramming/rprogramming_files/figure-html/nested_loops3-1.png" width="672" /></p>
</div>
</div>
<div id="exercise-4" class="section level3">
<h3>Exercise 4</h3>
<div id="build-your-own-r-package" class="section level4">
<h4>Build your own R package</h4>
<p><strong>Task 4.1</strong>: Save one or more of your functions to a file called <code>script.R</code> and build the package with the <code>package.skeleton</code> function.</p>
<pre class="r"><code>package.skeleton(name=&quot;mypackage&quot;, code_files=c(&quot;script1.R&quot;))</code></pre>
<p><strong>Task 4.2</strong>: Build tarball of the package</p>
<pre class="r"><code>system(&quot;R CMD build mypackage&quot;)</code></pre>
<p><strong>Task 4.3</strong>: Install and use package</p>
<pre class="r"><code>install.packages(&quot;mypackage_1.0.tar.gz&quot;, repos=NULL, type=&quot;source&quot;)
library(mypackage)
?myMAcomp # Opens help for function defined by mypackage</code></pre>
</div>
</div>
</div>
<div id="homework-5" class="section level2">
<h2>Homework 5</h2>
<p>See homework section <a href="https://girke.bioinformatics.ucr.edu/GEN242/assignments/homework/hw05/hw05/">here</a>.</p>
</div>
<div id="session-info" class="section level2">
<h2>Session Info</h2>
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>## R version 4.0.3 (2020-10-10)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Debian GNU/Linux 10 (buster)
## 
## Matrix products: default
## BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.8.0
## LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.8.0
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C               LC_TIME=en_US.UTF-8       
##  [4] LC_COLLATE=en_US.UTF-8     LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                  LC_ADDRESS=C              
## [10] LC_TELEPHONE=C             LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] stats     graphics  grDevices utils     datasets  methods   base     
## 
## other attached packages:
## [1] ggplot2_3.3.2    limma_3.46.0     BiocStyle_2.18.0
## 
## loaded via a namespace (and not attached):
##  [1] knitr_1.30          magrittr_2.0.1      tidyselect_1.1.0    munsell_0.5.0      
##  [5] colorspace_2.0-0    R6_2.5.0            rlang_0.4.8         dplyr_1.0.2        
##  [9] stringr_1.4.0       tools_4.0.3         grid_4.0.3          gtable_0.3.0       
## [13] xfun_0.20           withr_2.3.0         ellipsis_0.3.1      htmltools_0.5.1.1  
## [17] yaml_2.2.1          digest_0.6.27       tibble_3.0.4        lifecycle_0.2.0    
## [21] crayon_1.3.4        bookdown_0.21       purrr_0.3.4         BiocManager_1.30.10
## [25] codetools_0.2-16    vctrs_0.3.5         glue_1.4.2          evaluate_0.14      
## [29] rmarkdown_2.5       blogdown_1.1.7      stringi_1.5.3       pillar_1.4.7       
## [33] compiler_4.0.3      generics_0.1.0      scales_1.1.1        pkgconfig_2.0.3</code></pre>
</div>
<div id="references" class="section level2 unnumbered">
<h2>References</h2>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-Gentleman2008-xo" class="csl-entry">
Gentleman, Robert. 2008. <em><span>R</span> Programming for Bioinformatics (Chapman &amp; <span>Hall/CRC</span> Computer Science &amp; Data Analysis)</em>. 1 edition. Chapman; Hall/CRC. <a href="http://www.amazon.com/Programming-Bioinformatics-Chapman-Computer-Analysis/dp/1420063677">http://www.amazon.com/Programming-Bioinformatics-Chapman-Computer-Analysis/dp/1420063677</a>.
</div>
</div>
</div>
